src/Induction/NoInfiniteDescent.agda

```diff
@@ -0,0 +1,221 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- A standard consequence of accessibility/well-foundedness:
+-- the impossibility of 'infinite descent' from any (accessible)
+-- element x satisfying P to 'smaller' y also satisfying P
+------------------------------------------------------------------------
+
+{-# OPTIONS --cubical-compatible --safe #-}
+
+module Induction.NoInfiniteDescent where
+
+open import Data.Nat.Base as ℕ using (ℕ; zero; suc)
+open import Data.Nat.Properties as ℕ
+open import Data.Product.Base using (_,_; proj₁; ∃-syntax; _×_)
+open import Function.Base using (_∘_)
+open import Induction.WellFounded
+  using (WellFounded; Acc; acc; acc-inverse; module Some)
+open import Level using (Level)
+open import Relation.Binary.Core using (Rel)
+open import Relation.Binary.Construct.Closure.Transitive
+open import Relation.Binary.PropositionalEquality.Core
+open import Relation.Nullary.Negation.Core as Negation using (¬_)
+open import Relation.Unary
+  using (Pred; ∁; _∩_; _⇒_; Universal; IUniversal; Stable)
+
+private
+  variable
+    a r ℓ : Level
+    A : Set a
+
+------------------------------------------------------------------------
+-- Definitions
+
+module InfiniteDescent (_<_ : Rel A r) (P : Pred A ℓ)  where
+
+    DescentAt : Pred A _
+    DescentAt x = P x → ∃[ y ] y < x × P y
+
+    Acc⇒Descent : Set _
+    Acc⇒Descent = ∀[ Acc _<_ ⇒ DescentAt ]
+
+    Descent : Set _
+    Descent = ∀[ DescentAt ]
+
+    InfiniteDescendingSequence : (f : ℕ → A) → Set _
+    InfiniteDescendingSequence f = ∀ n → f (suc n) < f n
```

Comment:
This is unnecessarily parameterised on `P`?

Reply:
Aaargh, yes, you get me every time on this. I'm still stuck in the 1990s, when the LEGO and Coq sectioning mechanisms did 'auto-strengthening' of definitions to remove unused hypotheses. And of course, Agda modules don't work like that, and I clearly subconsciously still wish that they did!